### ISSUE TITLE
[QST] How to update Table.column value in libcudf Java API

### ISSUE THREAD

**Issue by hiroki-nishimoto-fixstars** [rapidsai/cudf#17652]
**What is your question?**

2 Questions. 

1. How can I write bellow cudf(Python) code to libcudf(Java) code?
2. Are there any introduction or documents or example codes for libcudf(Java)?

---

Hi. I'm trying to rewrite cudf(Python) code to libcudf(Java) code.

In my understanding, they are implemented diffrenct concept.  So, I do not believe that they are completely compatible, of course.

How can I write a process in libcudf(Java) that corresponds to the following process of updating a Series in a cudf.DataFrame?

libcudf(Java) prohibit to access `Table.columns` and don't have update column API. ex) `setColumn(index, value)` 

```python
import pandas as pd
df = pd.DataFrame({"a":[1,2,3], "b" : [4,5,6]})
df["a"] = [9,8,7]
print(df)


'''
output
  a b
0 9 4
1 8 5
2 7 6
'''
```

A simple implementation would yield the following code, which does not seem very smart.

```java
Table df = ... // omission

ColumnVector update_a = ColumnVector.fromInts(9,8,7);
int numCols = df.getNumberOfColumns();
ColumnVector cols = new ColumnVector[numCols];
for( int i = 0 ; i < numCols; i++ ){
    if( i != 0 ){
        cols[i] = df.getColumn[i];
    }
    else {
        cols[i] = update_a;
    }
}
df = new Table(cols);
```

Also, I know there is libcudf(C++) documentation, but I could not find any Java documentation or example code. Is there somewhere I can find it?

Now I'm trying to implement with  reading [Test code](https://github.com/rapidsai/cudf/tree/branch-25.02/java/src/test) and [API implementation](https://github.com/rapidsai/cudf/tree/branch-25.02/java/src/main)...

**Comment by GregoryKimball** [comment]
@ttnghia would you please share your thoughts on the best way to update a column using the libcudf Java API?

**Comment by ttnghia** [comment]
> libcudf(Java) prohibit to access Table.columns

No, cudf Java provides the function `getColumn(int index)` that allows accessing to the table columns and you already used it in your example. Your example above is almost correct. Look at the documentation of `Table.getColumn`:
```
   * .....If you want to keep a reference to
   * the column around past the life time of the table, you will need to increment the reference
   * count on the column yourself.
```
That means, this is needed when calling `getColumn`:
```
if( i != 0 ){
        cols[i] = df.getColumn[i];
        cols[i].incRefCount();
    }
```
Add that second line to your example. That's all you need.

**Comment by hiroki-nishimoto-fixstars** [comment]
Thank you for answering.

It's really helpful to know more about reference counting.

>  Your example above is almost correct. 

So, are the following conclusions correct?

- If I want to update one of column of table, I must declare new table with exist columns and new update column, and assign to exist table.
- Exist table column is not able to be updated directly.

**Comment by hiroki-nishimoto-fixstars** [comment]
Sorry for the late reply.

> A table is just a manager for the columns, so creating a new table holding existing columns of another table is relatively cheap.

With these words, I understand how best to update columns.

Thank you for your detailed reply.
